ZooKeeper 3.4.6 as a membership provider

shayhatsor opened the pull request:
You can now run Orleans with a zookeeper membership table:
```xml
<SystemStore
      DataConnectionString="192.168.1.1;192.168.1.2;192.168.1.3"
      DeploymentId="BestDeploymentEver"
      SystemStoreType="ZooKeeper"/>
```

For testing, please copy [ZooKeeper 3.4.6 release](http://www.apache.org/dyn/closer.cgi/zookeeper/) to some local folder. then go to "conf" folder and rename zoo_sample.cfg to zoo.cfg. then go to the bin folder and run zkServer.cmd. this runs the server at 127.0.0.1:2181, which is the default address for the tests.

While integrating I came across some issues:
my first instinct was to create an "OrleansZooKeeperUtils" project, but that created circular dependencies since IMembershipTable & IGatewayListProvider live in Orleans.dll. . so for now it's where the SqlMembershipTable, in the Orleans.dll. Also, the Orleans.csproj file contains hard coded references in the CodeGenToolReferences property, which i edited manually to allow compilation.

Also, my implementation doesn't include reminders. Which might be better implemented using a different DB, maybe MySQL.  As far as i can tell, ZooKeeper is **perfect** for implementing the membership table. but the reminders interface requires range queries, that aren't natural for ZooKeeper. So, for now, i had to fallback to grain in memory reminders(which i am aware is not a solution).

Obviously, this integration should be cleaner. now i'll need your help.

dnfclas commented:
Hi __@shayhatsor__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla2.dotnetfoundation.org.
    </span>

TTYL, DNFBOT;

dnfclas commented:
__@shayhatsor__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, DNFBOT;

gabikliot commented:
@shayhatsor , this is great! Thank you!
I am now on vacation, so I will provide a partial feedback now and will review the core of the ZooKeeperBasedMembershipTable.cs later on.

1) Regarding dependencies: the idea is not to introduce dependence on Zookeeper dll in Orleans. You can creeate OrleansZooKeeperUtils and load it dynamically, and access dynamically via reflection. You will basically load `ZooKeeperBasedMembershipTable` impl. of IMembership table in the factory (and GWProvider impl. of IGWProvider). That's it: since IMembership  interface is defined in Orleans and you won't be exposing any additional new types. Should be simple. We are doing the same now with dependencies on Azure (#136) - removing hard dependencies.

2) In the  GatewayProviderFactory it should also ZKGatewayListProvider and not static list. It should essentially just read all rows, and convert to the right format and return all silos. We decoupled IMembership  from IGatewayListProvider  to be able for example to return different sub list of silos to different clients. It allows more flexibility. But usually we just return the full list of silos.  

3) For reminders, it is OK for ZK case to use Grain for now. We can look at it later on, after MBR is done.

gabikliot commented on src/Orleans/Messaging/ZooKeeperBasedMembershipTable.cs:
```diff
@@ -0,0 +1,352 @@
+                try
+                {
+                    await Task.WhenAll(getRowTask, getTableNodeTask);
+                    rows.Add(getRowTask.Result);
```
better to write: rows.Add(await getRowTask);

gabikliot commented on src/Orleans/Messaging/ZooKeeperBasedMembershipTable.cs:
```diff
@@ -0,0 +1,352 @@
+                }
+                catch (KeeperException.NoNodeException)
+                {
+                    //that's ok
```
can you explain in the comment what is NoNodeException and why it is OK.

gabikliot commented on src/Orleans/Messaging/ZooKeeperBasedMembershipTable.cs:
```diff
@@ -0,0 +1,352 @@
+
+                var tableVersion = ConvertToTableVersion(childrenResult.Stat);
+
+                return new MembershipTableData(childrenTasks.Select(t => t.Result).ToList(), tableVersion);
```
await t instead of t.Result.

gabikliot commented on src/Orleans/Messaging/ZooKeeperBasedMembershipTable.cs:
```diff
@@ -0,0 +1,352 @@
+                    ipEndPointConverter, siloAddressConverter);
+        }
+
+        private class ZooKeeperWatcher : Watcher
```
what is the purpose of ZooKeeperWatcher ?

gabikliot commented on src/Orleans/Messaging/GatewayProviderFactory.cs:
```diff
@@ -50,6 +50,10 @@ internal class GatewayProviderFactory
                     break;
 
+                case ClientConfiguration.GatewayProviderType.ZooKeeper:
+                    listProvider = new StaticGatewayListProvider(cfg);
```
here should just create ZooKeeperBasedMembershipTable ?

sergeybykov commented on src/Orleans/packages.config:
```diff
@@ -1,10 +1,13 @@
   <package id="System.Spatial" version="5.6.0" targetFramework="net45" />
   <package id="WindowsAzure.Storage" version="4.2.0" targetFramework="net45" />
   <package id="WiX.Toolset" version="3.9.1006.0" targetFramework="net4000" />
+  <package id="ZooKeeperNetEx" version="3.4.6-alpha12" targetFramework="net45" />
```
We need to load this dependency dynamically. We are going to do the same for the dependency on Azure for the Azure Table-based membership.

sergeybykov commented on src/Orleans/packages.config:
```diff
@@ -1,10 +1,13 @@
   <package id="Microsoft.Data.OData" version="5.6.0" targetFramework="net45" />
   <package id="Microsoft.Data.Services.Client" version="5.6.0" targetFramework="net45" />
   <package id="Microsoft.WindowsAzure.ConfigurationManager" version="2.0.0" targetFramework="net45" />
+  <package id="Newtonsoft.Json" version="5.0.8" targetFramework="net45" />
```
Why not 6.0?

jthelin replied:
5.0.8 is the version we use everywhere else in Orleans code, so we should lock on using that one version everywhere.

sergeybykov commented on src/Orleans/packages.config:
```diff
@@ -1,10 +1,13 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
+  <package id="log4net" version="2.0.3" targetFramework="net45" />
```
Is this dependency really needed?

jthelin commented:
In general, I would much prefer if we had a separate assembly containing all the ZK related code (same way we do got storage providers) rather than adding lots of optional code into the core assemblies. Otherwise our deployment dependencies for core system multiply like crazy. 

On other words, have `ZooKeeperBasedMembershipTable.cs` in its one ZK-related project / DLL. 

We could change config subsystem to use a simple naming pattern to find and load specific membership provider classes.
For example: 
GatewayProviderType=`ZooKeeper` auto maps to `Orleans.Membership.ZooKeeper.dll` assembly and `Orleans.Membership.ZooKeeperBasedMembershipTable` class for dynamic load, or something similar.

gabikliot commented:
@jthelin, this has already been extensively discussed and agreed upon. For now we agreed to load ZK dll dynamically from membership factory class. It eliminates all static dependencies on  ZK dll. So essentially very close to what you have suggested above, even simpler.

jthelin commented:
> this has already been extensively discussed and agreed upon

Could you please point me to where that "decision" and the "extensive discussion" is documented publicly ? Seems I missed it. ;)

gabikliot commented:
@jthelin:
https://github.com/dotnet/orleans/issues/69#issuecomment-72387551
https://github.com/dotnet/orleans/issues/382#issuecomment-99640820 " Membership storage is already pluggable in the sources, as Gabi described. We could make it into a real plugin design with dynamically loaded implementations, but I think that's a low priority part at this point, as we don't expect too many of such implementations."

jthelin commented:
Well then, I strongly disagree that "dynamic load" is low-priority, due to the way that this ZK change bleeds extra dependencies into core runtime.

We need to start getting into the good habit of not changing "core runtime" when we add additional extension functionality like new "providers".

BTW, "dynamic loaded" and "name-based discovery" are complementary, but slightly orthogonal.

gabikliot commented:
@jthelin, can we please not confuse and slow down @shayhatsor ? How "dynamically" ZK is integrated in Orleans is something you, me and @sergeybykov can decide and agree later on. We can start with something simple, like what we said until now: MBR Factory load ZK.dll dynamically, and later potentially refactor that. 
The highest priority for me is not to block @shayhatsor in doing the actual work.

jthelin commented:
Latest updates LGTM. 
Thanks for working through the structural tweaks @shayhatsor 
We have clean dependencies again now :)

shayhatsor commented:
@gabikliot & @sergeybykov, thank you for you reviews.
please have a look at the corrected code. The main change is that I've added Initialize methods to *IMembershipTable* and *IGatewayListProvider*. It enables the plugin developer to implement custom async method that Orleans runs before any other method of the interface. It replaces the static factory methods which can't conform to an interface. Now it's easy to move any implementations of these interfaces out of the core projects to their own dlls. I am aware that this change separates construction and initialization, and objects are in an invalid state till their initialize method is called, but that seems OK because Orleans does that once and internally. What do you think ?

sergeybykov commented:
@shayhatsor I'm looking at the code now. It would really help if it was rebased and squashed to 2-3 commits. If you are more of a newbie to Git than even we are (I saw your 'git is hard for newbies' commit comment) :-) , we can help with that. This is a great contribution, and it would be nice to merge it as a couple of clean commits.
Thank you for doing the non-trivial work implementing it!

sergeybykov commented:
Looks good. And the cleanup and refactoring of the code around membership factory is nice.

shayhatsor commented:
@sergeybykov thanks !
I'll create the two clean commits (git won't beat me)
last thing, I've added a post build event to xcopy the OrleansZooKeeperUtils binaries to OrleansHost and TesterInternal. then I removed it because it didn't pass the remote build. the idea was to allow people to use orleans with zookeeper out of the box, without having to manually copy binaries around. what's the best practice in this case ?

sergeybykov commented:
@shayhatsor You can simply make TesterInternal reference OrleansZooKeeperUtils project. That should make build copy it, and I think it the right thing to do for testing.

I'm less sure about copying OrleansZooKeeperUtils for OrleansHost. I think we can leave that part for now, and figure it out later as we refactor OrleansAzureUtils to be dynamically loaded as well.

veikkoeeva commented:
> t replaces the static factory methods which can't conform to an interface. Now it's easy to move any implementations of these interfaces out of the core projects to their own dlls. I am aware that this change separates construction and initialization, and objects are in an invalid state till their initialize method is called, but that seems OK because Orleans does that once and internally.

@shayhatsor, @sergeybykov, @jthelin Have you noticed that the tests also rely on having ``internal Task DeleteMembershipTableEntries(string deploymentId)`` there which, I think, falls into the same category as the init methods?

I put my little testing hack in my PR branch, at least for the time being, and you can see [here](https://github.com/dotnet/orleans/pull/347/files#diff-efd64021c31fede872ca826ee79bc174R139) how I use ``dynamic`` to expose all the method. Before the "little hack" there were the concrete AzureTable implementation used, but the tests actually test the *membership protocol*, i.e. the ``IMembershipTable`` interface. As a side note, @shayhatsor, that is one way of also testing the ZooKeeper bits.

shayhatsor commented:
@veikkoeeva I've already did that in this PR. The init methods are the main change,  but I also added the delete method to IMembershipTable. the implementation would be incomplete without it.

gabikliot commented on src/OrleansRuntime/MembershipService/GrainBasedMembershipTable.cs:
```diff
@@ -32,6 +33,17 @@ internal class GrainBasedMembershipTable : Grain, IMembershipTable
 
+        public Task InitializeMembershipTable(GlobalConfiguration config, bool tryInitTableVersion, TraceLogger traceLogger)
+        {
+            return OnActivateAsync();
```
no need to call OnActivateAsync here. It will be called automatically before first msg is delivered, which happens to be InitializeMembershipTable. so can just move the logic from OnActivateAsync to here or InitializeMembershipTable empty. Otherwise, the init logic will execute twice.

shayhatsor replied:
ok, i think InitializeMembershipTable and DeleteMembershipTableEntries should just be:
```csharp
throw new InvalidOperationException("This method should not be called directly");
```
because it's there only as stubs and breaks the grain flow.

gabikliot commented:
@shayhatsor , the init logic looks OK now.
I actually already started doing something similar last week [here](https://github.com/gabikliot/orleans/commit/f90c7bbde3791bed33a055a2ca740d9073aad726), but haven't finished yet. Once your PR is in I will I merge mine with yours.

Overall looks good.
I want to do more testing and review the actual membership protocol more seriously. As I am out till next week, so the best I think would be to wait till early next week, if that is OK with you and everyone.

sergeybykov commented:
@gabikliot You want to test the ZK implementation of the protocol or the changes @shayhatsor made to the factory code? The former shouldn't break anybody. The latter should be covered by our internal testing. Am I missing something?

gabikliot commented:
Yes, I meant the protocol. Indeed, merging this as is seems to present no danger to the rest of the system. What ever comments I have to either the ZK protocol itself or even some tweaks to the integration, can be added later on. So can merge from my perspective.

sergeybykov commented:
@shayhatsor Ping me when you are done with rebasing and squashing. We are ready to test and merge.

shayhatsor commented:
@sergeybykov merge away

gabikliot commented on src/OrleansRuntime/MembershipService/GrainBasedMembershipTable.cs:
```diff
@@ -32,6 +34,16 @@ internal class GrainBasedMembershipTable : Grain, IMembershipTable
 
+        public Task InitializeMembershipTable(GlobalConfiguration config, bool tryInitTableVersion, TraceLogger traceLogger)
+        {
+            throw new InvalidOperationException("This method should not be called directly");
```
No, I think throwing is wrong here. It will break the common code in tests and all the usage of GrainBasedMembershipTable , since we are calling InitializeMembershipTable for all implementatiions.
I think it should be no-op instead.

shayhatsor replied:
my current code actually checks for that specifically:
```c
GlobalConfiguration.LivenessProviderType livenessType = config.LivenessType;
            if (livenessType.Equals(GlobalConfiguration.LivenessProviderType.MembershipTableGrain))
            {
                membershipTable =
                    GrainFactory.Cast<IMembershipTable>(GrainReference.FromGrainId(Constants.SystemMembershipTableId));
            }
            else
            {
```
i did that because calling InitializeMembershipTable on GrainBasedMembershipTable in the context of MembershipFactory throws 
```c
throw new InvalidExpressionException(
                    String.Format("Trying to send a message on a silo not from within grain and not from within system target (RuntimeContext is not set to SchedulingContext) "
                        + "RuntimeContext.Current={0} TaskScheduler.Current={1}",
                        RuntimeContext.Current == null ? "null" : RuntimeContext.Current.ToString(),
                        TaskScheduler.Current));
```
anyway, to preserve consistency, maybe this is a better solution:
```c
public Task InitializeMembershipTable(GlobalConfiguration config, bool tryInitTableVersion, TraceLogger traceLogger)
        {
            return TaskDone.Done;
        }

        public Task DeleteMembershipTableEntries(string deploymentId)
        {
            table = new InMemoryMembershipTable();
            return TaskDone.Done;
        }
```

also, should DeleteMembershipTableEntries delete all rows but the deployment row itself? (the ZK implementation currently deletes the deployment itself with all rows). I think that it should be named DeleteMembershipTable and get a bool whether to delete the deployment itself or not. this will make the method more testable and easy to understand.

sergeybykov commented:
@shayhatsor Thank you! This was a major contribution. Now Orleans has a highly available cluster membership option other than Azure Table.
